components/script/fetch.rs

```diff
@@ -266,6 +273,201 @@ pub(crate) fn Fetch(
     promise
 }
 
+/// <https://fetch.spec.whatwg.org/#queue-a-deferred-fetch>
+fn queue_deferred_fetch(
+    request: NetTraitsRequest,
+    activate_after: Finite<f64>,
+    global: &GlobalScope,
+) -> Arc<Mutex<DeferredFetchRecord>> {
+    let trusted_global = Trusted::new(global);
+    // Step 1. Populate request from client given request.
+    // TODO
+    // Step 2. Set request’s service-workers mode to "none".
+    // TODO
+    // Step 3. Set request’s keepalive to true.
+    // TODO
+    // Step 4. Let deferredRecord be a new deferred fetch record whose request is request, and whose notify invoked is onActivatedWithoutTermination.
+    let deferred_record = Arc::new(Mutex::new(DeferredFetchRecord {
+        request,
+        global: trusted_global.clone(),
+        invoke_state: Cell::new(DeferredFetchRecordInvokeState::Pending),
+        activated: Cell::new(false),
+    }));
+    // Step 5. Append deferredRecord to request’s client’s fetch group’s deferred fetch records.
+    // TODO
+    // Step 6. If activateAfter is non-null, then run the following steps in parallel:
+    let now = Instant::now();
+    let deferred_record_clone = deferred_record.clone();
+    global.task_manager().deferred_fetch_task_source().queue(
+        task!(process_deferred_record: move || {
+            // Step 6.1. The user agent should wait until any of the following conditions is met:
+            let elapsed = now.elapsed().as_millis() as u64;
+            let activate_after = *activate_after as u64;
+            if elapsed < activate_after {
+                let sleep_duration = activate_after - elapsed;
+                thread::sleep(Duration::from_millis(sleep_duration));
```

Comment:
I looked for existing usages of running "in parallel" and how to deal with timers. I spotted this code in https://github.com/servo/servo/blob/d12dc23083c140f3e05e4c49e1cb443fc7d836c6/components/webdriver_server/actions.rs#L169-L174 and it appears to work. Not sure if this is correct though, as I think we block the main thread with this? In any case, the tests all appear to be happy, so maybe it is okay after all.